### PR TITLE
apply tree solution

### PR DESCRIPTION
**Pull request by VagoLali**
image

### CODE REVIEW

File: src/pricer.py [hunk 1]
```diff
@@ -337,40 +337,10 @@ def calc_rho(self, method: GreekMethod) -> float:
...
-                price_tree[step][i] = discounted_price
-        return price_tree[0][0]
-
-class AmericanTreePricer(Pricer):
-    def __init__(self, contract: AmericanContract, model: MarketModel, params: TreeParams):
-        if not isinstance(contract, AmericanContract):
-            raise TypeError(f'Contract must be of type AmericanContract but received {type(contract).__name__}')
+class TreePricer(Pricer):
+    def __init__(self, contract: Contract, model: MarketModel, params: TreeParams):
+        # if not isinstance(contract, EuropeanContract):
```
Comment (friedgab) [inline src/pricer.py:342]: Could you please remove these comments?

File: src/pricer.py [hunk 2]
```diff
@@ -380,6 +350,9 @@ def __init__(self, contract: AmericanContract, model: MarketModel, params: TreeP
             tree_method = SimpleBinomialTree(self.contract, self.model, self.params)
         self.tree_method = tree_method
 
+    def pre_final_value(self, spot: dict[float, float], step: int, discounted_continuation_value: float) -> float:
+        raise RuntimeError('pre_final_value must be overriden in child classes.')
```
Comment (friedgab) [inline src/pricer.py:354]: Remove the error, inherit the class from ABC and mark the this method as abstractmethod.

**Review by friedgab — CHANGES_REQUESTED** [review]

File: src/pricer.py (hunk 1 above)
Comment (VagoLali) [reply src/pricer.py]: done

File: src/pricer.py (hunk 2 above)
Comment (VagoLali) [reply src/pricer.py]: did it, unit tests run, does it look good?

**Review by friedgab — APPROVED** [review]
Looks good. Thanks.

**Pull request merged**